pkg/clients/azure/aks.go

```diff
@@ -1,203 +0,0 @@
-type AKSClusterAPI interface {
-	Get(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.ManagedCluster, error)
-	CreateOrUpdateBegin(ctx context.Context, instance computev1alpha1.AKSCluster, clusterName, appID, spSecret string) ([]byte, error)
-	CreateOrUpdateEnd(op []byte) (bool, error)
-	Delete(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.ManagedClustersDeleteFuture, error)
-	ListClusterAdminCredentials(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.CredentialResults, error)
-}
-
-// AKSClusterClient is the concreate implementation of the AKSClusterAPI interface that calls Azure API.
-type AKSClusterClient struct {
-	containerservice.ManagedClustersClient
-}
-
-// NewAKSClusterClient creates and initializes a AKSClusterClient instance.
-func NewAKSClusterClient(provider *v1alpha1.Provider, clientset kubernetes.Interface) (*AKSClusterClient, error) {
-	client, err := NewClient(provider, clientset)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create Azure client: %+v", err)
-	}
-
-	aksClustersClient := containerservice.NewManagedClustersClient(client.SubscriptionID)
-	aksClustersClient.Authorizer = client.Authorizer
-	aksClustersClient.AddToUserAgent(UserAgent)
-
-	return &AKSClusterClient{aksClustersClient}, nil
-}
-
-// Get returns the AKS cluster details for the given instance
-func (c *AKSClusterClient) Get(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.ManagedCluster, error) {
-	return c.ManagedClustersClient.Get(ctx, instance.Spec.ResourceGroupName, instance.Status.ClusterName)
-}
-
-// CreateOrUpdateBegin begins the create/update operation for a AKS Cluster with the given properties
-func (c *AKSClusterClient) CreateOrUpdateBegin(ctx context.Context, instance computev1alpha1.AKSCluster, clusterName, appID, spSecret string) ([]byte, error) {
-	spec := instance.Spec
-
-	agentPoolProfileName := fmt.Sprintf(AgentPoolProfileNameFmt, clusterName)
-	enableRBAC := !spec.DisableRBAC
-
-	nodeCount := int32(computev1alpha1.DefaultNodeCount)
-	if spec.NodeCount != nil {
-		nodeCount = int32(*spec.NodeCount)
-	}
-
-	createParams := containerservice.ManagedCluster{
-		Name:     &clusterName,
-		Location: &spec.Location,
-		ManagedClusterProperties: &containerservice.ManagedClusterProperties{
-			KubernetesVersion: &spec.Version,
-			DNSPrefix:         &spec.DNSNamePrefix,
-			AgentPoolProfiles: &[]containerservice.ManagedClusterAgentPoolProfile{
-				{
-					Name:   &agentPoolProfileName,
-					Count:  &nodeCount,
-					VMSize: containerservice.VMSizeTypes(spec.NodeVMSize),
-				},
-			},
-			ServicePrincipalProfile: &containerservice.ManagedClusterServicePrincipalProfile{
-				ClientID: to.StringPtr(appID),
-				Secret:   to.StringPtr(spSecret),
-			},
-			EnableRBAC: &enableRBAC,
-		},
-	}
-
-	createFuture, err := c.CreateOrUpdate(ctx, instance.Spec.ResourceGroupName, clusterName, createParams)
```

Comment:
this completely removes the tracking ability for the create request/operation.  we won't be able to check when it's done or get the root cause error if we don't keep this future object.

Reply:
For status tracking, we are using managed cluster `ProvisioningState` which is IMO correctly reflects current status, i.e. we do not need `Future` object fo that. 

I agree with you that we still need `Future` if we want more details in case of the `Failure`. Which seems to be the only reason why we need it. 

What do you think?